// tween-one/lib/index.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
var _exportNames = {
  Plugins: true,
  StylePlugin: true,
  ArrayPlugin: true,
  GradientsPlugin: true,
  PathMotionPlugin: true,
  SvgDrawPlugin: true,
  SvgMorphPlugin: true,
  Ticker: true,
  Easing: true,
  Tween: true,
  globalTimeline: true
};
Object.defineProperty(exports, "ArrayPlugin", {
  enumerable: true,
  get: function get() {
    return _ArrayPlugin.default;
  }
});
Object.defineProperty(exports, "Easing", {
  enumerable: true,
  get: function get() {
    return _ease.default;
  }
});
Object.defineProperty(exports, "GradientsPlugin", {
  enumerable: true,
  get: function get() {
    return _GradientsPlugin.default;
  }
});
Object.defineProperty(exports, "PathMotionPlugin", {
  enumerable: true,
  get: function get() {
    return _PathMotionPlugin.default;
  }
});
Object.defineProperty(exports, "Plugins", {
  enumerable: true,
  get: function get() {
    return _plugins.default;
  }
});
Object.defineProperty(exports, "StylePlugin", {
  enumerable: true,
  get: function get() {
    return _StylePlugin.default;
  }
});
Object.defineProperty(exports, "SvgDrawPlugin", {
  enumerable: true,
  get: function get() {
    return _SvgDrawPlugin.default;
  }
});
Object.defineProperty(exports, "SvgMorphPlugin", {
  enumerable: true,
  get: function get() {
    return _SvgMorphPlugin.default;
  }
});
Object.defineProperty(exports, "Ticker", {
  enumerable: true,
  get: function get() {
    return _ticker.default;
  }
});
Object.defineProperty(exports, "Tween", {
  enumerable: true,
  get: function get() {
    return _TweenOne.Tween;
  }
});
exports.default = void 0;
Object.defineProperty(exports, "globalTimeline", {
  enumerable: true,
  get: function get() {
    return _TweenOne.globalTimeline;
  }
});

var _plugins = _interopRequireDefault(require("./plugins"));

var _StylePlugin = _interopRequireDefault(require("./plugins/StylePlugin"));

var _ArrayPlugin = _interopRequireDefault(require("./plugins/ArrayPlugin"));

var _GradientsPlugin = _interopRequireDefault(require("./plugins/GradientsPlugin"));

var _PathMotionPlugin = _interopRequireDefault(require("./plugins/PathMotionPlugin"));

var _SvgDrawPlugin = _interopRequireDefault(require("./plugins/SvgDrawPlugin"));

var _SvgMorphPlugin = _interopRequireDefault(require("./plugins/SvgMorphPlugin"));

var _ticker = _interopRequireDefault(require("./animate/ticker"));

var _ease = _interopRequireDefault(require("./animate/ease"));

var _TweenOne = _interopRequireWildcard(require("./TweenOne"));

var _typings = require("./typings");

Object.keys(_typings).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _typings[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _typings[key];
    }
  });
});

var _IAnimObject = require("./typings/IAnimObject");

Object.keys(_IAnimObject).forEach(function (key) {
  if (key === "default" || key === "__esModule") return;
  if (Object.prototype.hasOwnProperty.call(_exportNames, key)) return;
  if (key in exports && exports[key] === _IAnimObject[key]) return;
  Object.defineProperty(exports, key, {
    enumerable: true,
    get: function get() {
      return _IAnimObject[key];
    }
  });
});
var _default = _TweenOne.default;
exports.default = _default;
